scylla/src/frame/value.rs

```diff
@@ -410,6 +418,82 @@ impl Value for Counter {
     }
 }
 
+fn zig_zag_encode(v: i64) -> u64 {
+    ((v >> 63) ^ (v << 1)) as u64
+}
+
+fn zig_zag_decode(v: u64) -> i64 {
+    ((v >> 1) as i64) ^ -((v & 1) as i64)
+}
+
+fn vint_encode(v: i64, buf: &mut Vec<u8>) {
+    let mut v = zig_zag_encode(v);
+    let number_of_bytes = 8 - v.leading_zeros() / 8;
+
+    if number_of_bytes != 8 {
+        let mut extra_bytes = number_of_bytes - 1;
+        if number_of_bytes > v.leading_zeros() % 8 {
+            extra_bytes += 1;
+        }
+        v += (!((0xff >> extra_bytes) as u8) as u64) << (8 * extra_bytes);
+    } else {
+        buf.put_u8(0xff);
+    }
+
+    buf.put_uint(v, number_of_bytes as usize);
+}
+
+pub(crate) fn vint_decode(buf: &[u8]) -> (i64, usize) {
+    let extra_bytes = buf[0].leading_ones() as usize;
+    let mut v = 0;
+
+    for (i, byte) in buf.iter().enumerate().take(extra_bytes + 1).skip(1) {
+        v += (*byte as u64) << (8 * (extra_bytes - i) as u64);
+    }
+
+    if extra_bytes != 8 {
+        v += ((buf[0] - !((0xff >> extra_bytes) as u8)) as u64) << (8 * extra_bytes);
+    }
+
+    (zig_zag_decode(v), extra_bytes + 1)
+}
+
+#[test]
+fn zig_zag_encode_test() {
+    assert_eq!(zig_zag_encode(0), 0);
+    assert_eq!(zig_zag_encode(-1), 1);
+    assert_eq!(zig_zag_encode(1), 2);
+    assert_eq!(zig_zag_encode(-2), 3);
+    assert_eq!(zig_zag_encode(2), 4);
+    assert_eq!(zig_zag_encode(-3), 5);
+    assert_eq!(zig_zag_encode(3), 6);
+}
+
+#[test]
+fn vint_encode_test() {
+    let mut buf: Vec<u8> = Vec::with_capacity(128);
+    let n: i64 = 256000;
+    vint_encode(n, &mut buf);
+    assert_eq!(buf, vec![199, 208, 0])
```

Comment:
I think we need more comprehensive tests for (unsigned) vint serialization and deserialization:

- Do the above check with some more numbers with different lengths,
  - Include corner cases - 0 and maxint,
- For each bit length of a number, choose a number of such length and check if serialization then deserialization returns the original number.